consensus/consensus-types/src/common.rs

```diff
@@ -33,29 +41,77 @@ pub struct RejectedTransactionSummary {
     pub hash: HashValue,
 }
 
+#[derive(Debug)]
+pub enum DataStatus {
+    Cached(Vec<SignedTransaction>),
+    Requested(Vec<oneshot::Receiver<Result<Vec<SignedTransaction>, Error>>>),
+}
+
+// impl PartialEq for DataStatus {
+//     fn eq(&self, other: &Self) -> bool {
+//         match (self, other) {
+//             (DataStatus::Cached(ltxn), DataStatus::Cached(rtxn)) => ltxn == rtxn,
+//             (DataStatus::Requested(lchannels), DataStatus::Requested(rchannels)) => Arc::as_ptr(&lchannels) == Arc::as_ptr(&rchannels),
+//             (_, _) => false,
+//         }
+//     }
+// }
+//
+// impl Eq for DataStatus {}
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+pub struct ProofWithStatus {
+    pub proofs: Vec<ProofOfStore>,
+    #[serde(skip)]
+    pub status: Arc<Mutex<Option<DataStatus>>>,
+}
+
+impl PartialEq for ProofWithStatus {
+    fn eq(&self, other: &Self) -> bool {
+        self.proofs == other.proofs && Arc::as_ptr(&self.status) == Arc::as_ptr(&other.status)
+    }
+}
+
+impl Eq for ProofWithStatus {}
+
 /// The payload in block.
 #[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq)]
 pub enum Payload {
     DirectMempool(Vec<SignedTransaction>),
+    InQuorumStore(ProofWithStatus),
+    Empty,
 }
 
 impl Payload {
     pub fn empty() -> Self {
-        Payload::DirectMempool(Vec::new())
+        Payload::Empty
```

Comment:
We can add `Empty` to the enum right now, but we just can't create it yet because that will break wire compatibility during a binary upgrade. So can we make this return `Payload::DirectMempool(Vec::new())` and replace the call of `Payload::Empty` to `Payload::empty()`? Then when the binary is released, we can make this return `Payload::Empty`.

Reply:
I think this will work. 
The other option is:
if QS enabled {
   Payload::InQuorumStore(Vec::new())
   } else{
Payload::DirectMempool(Vec::new())
}

This will allow us to get rid of Empty.
If we propagate this information to the round manager in order to vote we might as well do the above.
What do you think is better?